catalog/ui/catalog-ui-search/src/main/java/org/codice/ddf/catalog/ui/query/suggestion/MgrsCoordinateProcessor.java

```diff
@@ -0,0 +1,340 @@
+   */
+  @Nullable
+  private LiteralSuggestion getMgrsSuggestion(String query) {
+    final List<String> matches = getMgrsCoordinateStrings(query);
+    if (matches.isEmpty()) {
+      LOGGER.debug("No valid MGRS strings were found or could be inferred from query [{}]", query);
+      return null;
+    }
+
+    final List<UsngCoordinate> mgrsCoordinates = new ArrayList<>();
+    final StringBuilder nameBuilder = new StringBuilder("MGRS: ");
+    for (String match : matches) {
+      UsngCoordinate coordinate = parseMgrsString(match);
+      if (coordinate != null) {
+        mgrsCoordinates.add(coordinate);
+        nameBuilder.append("[ ").append(coordinate.toMgrsString()).append(" ] ");
+      }
+    }
+
+    if (mgrsCoordinates.size() == 1) {
+      final UsngCoordinate mgrsCoord = mgrsCoordinates.get(0);
+      final BoundingBox bb = translator.toBoundingBox(mgrsCoord);
+      return new LiteralSuggestion(
+          LITERAL_SUGGESTION_ID, nameBuilder.toString(), latLonFromBoundingBox(bb));
+    }
+    return new LiteralSuggestion(
+        LITERAL_SUGGESTION_ID,
+        nameBuilder.toString(),
+        mgrsCoordinates
+            .stream()
+            .map(translator::toLatLon)
+            .map(d -> new LatLon(d.getLat(), d.getLon()))
+            .collect(Collectors.toList()));
+  }
+
+  @Nullable
+  private UsngCoordinate parseMgrsString(String mgrs) {
+    try {
+      return translator.parseMgrsString(mgrs);
+    } catch (ParseException e) {
+      LOGGER.debug("Detected string [{}] was not valid MGRS", mgrs, e);
+      return null;
+    }
+  }
+
+  private static List<LatLon> latLonFromBoundingBox(BoundingBox bb) {
+    return ImmutableList.of(
+        new LatLon(bb.getSouth(), bb.getWest()),
+        new LatLon(bb.getNorth(), bb.getWest()),
+        new LatLon(bb.getNorth(), bb.getEast()),
+        new LatLon(bb.getSouth(), bb.getEast()));
+  }
+
+  /**
+   * Given the user's query, return a list of valid MGRS strings that could be inferred from the
+   * query. The strings should be ready for submission to usng4j to be parsed and mathematically
+   * evaluated.
+   *
+   * <p>The query is terminated with whitespace so the user does not have to. See the LHS of {@link
+   * #PATTERN_MGRS_NUMERIC} for the cases where this is useful.
+   *
+   * <p>There should be no duplicate entries in the returned list. The order of the entries matters
+   * so a {@link java.util.Set} cannot be used.
+   *
+   * @param query the input string potentially holding MGRS values.
+   * @return a list of valid MGRS values that can be sent to usng4j.
+   */
+  @VisibleForTesting
+  List<String> getMgrsCoordinateStrings(String query) {
+    final String terminatedQuery = query.concat(WHITESPACE);
+    final Matcher matcher = PATTERN_MGRS_ANYPART.matcher(terminatedQuery);
+    final List<String> mgrsStrings = new ArrayList<>();
+
+    StringBuilder mgrsBuilder = new StringBuilder();
+    MgrsPartType previousType = MgrsPartType.NONE;
```

Comment:
:exclamation: :question: Where is previousType mutated?

Reply:
I update `previousType` to be `type` at `213` and I'm using it in the loop so I know my previous state.